src/BinaryPack/Extensions/System/TypeExtensions.cs

```diff
@@ -56,17 +55,65 @@ orderby member.Name
                       (mode.HasFlag(SerializationMode.NonPublicMembers) || member.IsPublic() ||
                        member.IsDefined(typeof(SerializableMemberAttribute))) &&
                       (member is FieldInfo fieldInfo && !fieldInfo.IsInitOnly && !fieldInfo.IsLiteral ||
-                       member is PropertyInfo propertyInfo && propertyInfo.CanRead && propertyInfo.CanWrite
+                       member is PropertyInfo propertyInfo && propertyInfo.CanRead
                        && propertyInfo.GetIndexParameters().Length == 0)
-                orderby member.Name
                 select member
-            }).ToArray();
+            }).SortMemberAsConstructorParameters(type)
+              .ToArray();
 
             return members;
         }
 
+        
+        /// Finding a constructor where all serialized members have a parameter with same type and same name (ignoring case)
+        private static IEnumerable<MemberInfo> SortMemberAsConstructorParameters(this IEnumerable<MemberInfo> members, Type type)
+        {
+            // Checking for constructor with no parameter and even any constructors first
+            if (type.GetConstructor(Type.EmptyTypes) != null || type.GetConstructors().Length == 0)
+            {
+                return members.OrderBy(member => member.Name);
+            }
+            
+            // Caching to prevent multiple iterations.
+            IEnumerable<MemberInfo> memberInfos = members as MemberInfo[] ?? members.ToArray();
+
+            List<MemberInfo> orderedMembers = new(memberInfos.Count());
+            
+            foreach (ConstructorInfo constructor in type.GetConstructors())
+            {
+                if (constructor.GetParameters().Length != memberInfos.Count())
+                {
+                    continue;
+                }
+                
+                orderedMembers.Clear();
+                bool foundAllParameters = true;
+
+                foreach (ParameterInfo parameter in constructor.GetParameters())
+                {
+                    MemberInfo? memberInfoForParameter = memberInfos.FirstOrDefault(x => x.GetMemberType() == parameter.ParameterType &&
+                                                                                         string.Equals(x.Name, parameter.Name, StringComparison.OrdinalIgnoreCase));
+
+                    if (memberInfoForParameter == null)
+                    {
+                        foundAllParameters = false;
+                        break;
+                    }
+                    
+                    orderedMembers.Add(memberInfoForParameter);
+                }
+
+                if (foundAllParameters)
+                {
+                    return orderedMembers;
+                }
+            }
+            
+            throw new NullReferenceException($"The type {type} has no constructor with parameters named like all Members.");
```

Comment:
NullReferenceException is misleading here. Something like InvalidOperationException would be more appropriate for the issue.